File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.py
class Solution(object):
    def checkValidString(self, s):
        leftmin, leftmax = 0, 0
        for c in s:
            if c == '(':
                leftmin += 1
                leftmax += 1
            elif c == ')':
                leftmin -= 1
                leftmax -= 1
            else:
                leftmin -= 1
                leftmax += 1
            if leftmax < 0:
                return False
            if leftmin < 0:
                leftmin = 0
            
        return leftmin == 0
        

File: 0678-valid-parenthesis-string/0678-valid-parenthesis-string.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int leftmin = 0, leftmax = 0;
        for(char c : s) {
            if(c == '(') leftmin++, leftmax++;
            else if(c == ')') leftmin--, leftmax--;
            else leftmax++, leftmin--;
            if(leftmax < 0) return false;
            if(leftmin < 0) leftmin = 0;
        }

        return leftmin == 0;
    }
};